### PR TITLE
KAFKA-16770; [1/N] Coalesce records into bigger batches

### PR DESCRIPTION
**Pull request by dajac**
We have discovered during large scale performance tests that the current write path of the new coordinator does not scale well. The issue is that each write operation writes synchronously from the coordinator threads. Coalescing records into bigger batches helps drastically because it amortizes the cost of writes. Aligning the batches with the snapshots of the timelines data structures also reduces the number of in-flight snapshots.

This patch is the first of a series of patches that will bring records coalescing into the coordinator runtime. As a first step, we had to rework the PartitionWriter interface and move the logic to build MemoryRecords from it to the CoordinatorRuntime. The main changes are in these two classes. The others are related mechanical changes.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: group-coordinator/src/test/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntimeTest.java [hunk 1]
```diff
@@ -1072,8 +1174,8 @@ public void replay(
     @Test
     public void testScheduleWriteOpWhenWriteFails() {
         MockTimer timer = new MockTimer();
-        // The partition writer only accept on write.
-        MockPartitionWriter writer = new MockPartitionWriter(2);
+        // The partition writer only accept one write.
```
Comment (jolshan) [inline group-coordinator/src/test/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntimeTest.java:1177]: for my understanding, we always batched the (in this case 2) records that were part of the same write operation. For now we aren't changing this, but moving the logic to the coordinator runtime to make space for the batching logic as a followup?

**Comment by jolshan** [comment]
I took a first pass to get a general understanding. I will come back tomorrow and take a deeper dive in some of the minor changes and let you know if i think of anything missed.

File: group-coordinator/src/test/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntimeTest.java (hunk 1 above)
Comment (dajac) [reply group-coordinator/src/test/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntimeTest.java:1177]: You got it right. A write operation produces a single batch with all the records generated by it. This patch does not change it but change where the memory record is built. The next patch will add the logic to keep the batch open until full or until a linger time is reached. With this, records produced by many write operations will end up in the same batch.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java [hunk 2]
```diff
@@ -1046,12 +1118,18 @@ public void run() {
...
-                        long offset = partitionWriter.appendEndTransactionMarker(
+                        long offset = partitionWriter.append(
                             tp,
-                            producerId,
-                            producerEpoch,
-                            coordinatorEpoch,
-                            result
+                            VerificationGuard.SENTINEL,
+                            MemoryRecords.withEndTransactionMarker(
+                                time.milliseconds(),
```
Comment (jolshan) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java:1125]: It seems we didn't specify this time value before. Was that a bug? I guess it also just gets the system time in the method.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java [hunk 3]
```diff
@@ -723,30 +757,66 @@ public void run() {
                         // If the records are not empty, first, they are applied to the state machine,
                         // second, then are written to the partition/log, and finally, the response
                         // is put into the deferred event queue.
+                        long prevLastWrittenOffset = context.coordinator.lastWrittenOffset();
+                        LogConfig logConfig = partitionWriter.config(tp);
+                        byte magic = logConfig.recordVersion().value;
+                        int maxBatchSize = logConfig.maxMessageSize();
+                        long currentTimeMs = time.milliseconds();
+                        ByteBuffer buffer = context.bufferSupplier.get(Math.min(16384, maxBatchSize));
```
Comment (jolshan) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java:765]: Nice we got rid of the thread local. 👍

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java [hunk 4]
```diff
@@ -723,30 +757,66 @@ public void run() {
...
+
                         try {
-                            // Apply the records to the state machine.
-                            if (result.replayRecords()) {
-                                // We compute the offset of the record based on the last written offset. The
-                                // coordinator is the single writer to the underlying partition so we can
-                                // deduce it like this.
-                                for (int i = 0; i < result.records().size(); i++) {
+                            MemoryRecordsBuilder builder = MemoryRecords.builder(
```
Comment (jolshan) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java:768]: nit: is there a benefit from putting this here and not right before the append method?

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/PartitionWriter.java [hunk 5]
```diff
@@ -18,40 +18,21 @@
...
-         * Serializes the value of the record.
-         */
-        byte[] serializeValue(T record);
-    }
+public interface PartitionWriter {
 
     /**
      * Listener allowing to listen to high watermark changes. This is meant
-     * to be used in conjunction with {{@link PartitionWriter#append(TopicPartition, List)}}.
+     * to be used in conjunction with {{@link PartitionWriter#append(TopicPartition, VerificationGuard, MemoryRecords)}}.
```
Comment (jolshan) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/PartitionWriter.java:35]: Is there a programatic way to check if these links are broken due to refactoring, or do you need to do it manually?

Just wondering if there is an easy way to check you did them all :)

File: group-coordinator/src/test/java/org/apache/kafka/coordinator/group/runtime/InMemoryPartitionWriter.java [hunk 6]
```diff
@@ -18,162 +18,29 @@
...
+import java.util.stream.StreamSupport;
 
 /**
  * An in-memory partition writer.
- *
- * @param <T> The record type.
  */
-public class InMemoryPartitionWriter<T> implements PartitionWriter<T> {
-
-    public static class LogEntry {
```
Comment (jolshan) [inline group-coordinator/src/test/java/org/apache/kafka/coordinator/group/runtime/InMemoryPartitionWriter.java:42]: nice that we could just use the real memory records

File: core/src/test/scala/unit/kafka/coordinator/group/CoordinatorPartitionWriterTest.scala [hunk 7]
```diff
@@ -84,98 +67,28 @@ class CoordinatorPartitionWriterTest {
   }
 
   @Test
-  def testWriteRecords(): Unit = {
```
Comment (jolshan) [inline core/src/test/scala/unit/kafka/coordinator/group/CoordinatorPartitionWriterTest.scala:87]: Do we have an equivalent test for the writing of the records in CoordinatorRuntimeTest? I didn't really notice new tests, but saw we have some of the builder logic there. Is it tested by checking equality between the records generated by the helper methods and the output from running the CoordinatorRuntime code?

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java (hunk 2 above)
Comment (dajac) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java:1125]: `withEndTransactionMarker` takes the current time if we don't specify it. The reason why I set it explicitly here is to ensure that the mock time is used in tests.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java (hunk 4 above)
Comment (dajac) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java:768]: The `builder` is used in the above loop (L801) so we need it here.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/PartitionWriter.java (hunk 5 above)
Comment (dajac) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/PartitionWriter.java:35]: Intellij reports them as warning. I suppose that we would get warning when we generate the javadoc too.

File: core/src/test/scala/unit/kafka/coordinator/group/CoordinatorPartitionWriterTest.scala (hunk 7 above)
Comment (dajac) [reply core/src/test/scala/unit/kafka/coordinator/group/CoordinatorPartitionWriterTest.scala:87]: Right. We have many tests in CoordinatorRuntimeTest doing writes. As we fully validate the records now, they cover this.

**Comment by dajac** [comment]
@jolshan Thanks for your comments. I replied to them.

**Review by jolshan — APPROVED** [review]

**Pull request merged**